### DISCUSSION TITLE
Expose `inputParser` on `Procedure`?

### DISCUSSION THREAD

**Discussion by mmkal** [trpc/trpc#714] [Ideas]
Hi - have just started using trpc and it is 👌 

One request: what do you think about exposing the (zod) schema on the `Procedure` object? The use-case being, I would like to create a UI in an internal admin app using [react-json-schema-form](https://github.com/rjsf-team/react-jsonschema-form) for various whitelisted "admin" queries/mutations. If we could access `appRouter._def.queries.myQuery.inputParser` we could pass it to [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema) which would give us some schemas we could use with react-json-schema-form.

It's possible to do now by suppressing the typescript error about using a private field, but I wanted to see if you'd be open to making it public so that it wouldn't disappear and break our workflow in a minor update.

**Comment by KATT** [comment]
On phone, will give example tomorrow, but

1. Define schema in React component
2. Import said schema into your the trpc router

**Comment by mmkal** [comment]
Interested to see the example, but I think I may not have explained the use case properly. I want to take an appRouter defined on the server side, with `@trpc/server` and zod inputs, and use it to get json-schema POJOs. Those POJOs can be used as the model for a json-schema based form in [retool](https://retool.com/) which we already use to build admin UIs. Retool happens to support [json-schema forms](https://docs.retool.com/docs/working-with-json-schema-form) - we just need to serve the schema POJO(s) for its model. We have done this in the past with AWS lambdas which validate their inputs with zod too.

So given that (desired) workflow I'm not your suggestion applies - the schema has already been defined on the server side, and there isn't actually a react app to move the schema definition to even if we wanted to.

As mentioned this is possible already by just accessing the private variable, which works at runtime, and the compiler error can be suppressed. Likely the pull request to fix this would just be removing `private` from the `inputParser`.

**Comment by KATT** [comment]
Ohhh, using trpc with Retool is exciting, I'd love for that to work well.

I'll have a peek later today and see if we can find a solution.

**Comment by KATT** [comment]
This is using the private property `Procedure::inputParser` so it's a bit of an unstable hack but it works, see b3427be - accessible at http://localhost:3000/api/trpc/posts.edit.input

<img width="542" alt="Screenshot 2021-07-23 at 12 55 33" src="https://user-images.githubusercontent.com/459267/126772508-22b80709-03e1-42af-a3be-257b84af19f7.png">

**Comment by KATT** [comment]
You can use the above, just be aware it may stop working in the future without a major version change.

If you can make a proof of concept and document it in the official docs I would happily add better support.

**Comment by KATT** [comment]
@all-contributors add @mmkal for ideas

**Comment by allcontributors** [comment]
@KATT 

I've put up [a pull request](https://github.com/trpc/trpc/pull/704) to add @mmkal! :tada:

**Comment by KATT** [comment]
did an example that does it dynamically in 43f328c
<img width="603" alt="Screenshot 2021-07-23 at 13 31 21" src="https://user-images.githubusercontent.com/459267/126775941-4477aa96-724d-4ad6-875e-f62f144b8f80.png">

**Comment by mmkal** [comment]
Amazing, thank you! Yes that example is exactly along the lines I was thinking. I will try to open a docs PR and make it non-private soon if I can get some spare cycles.

**Comment by KATT** [comment]
@mmkal - I'll leave the PR + docs to you then. Would be really cool get an example of this up and running on retool, such a timesaver.
